mlflow/autogen/autogen_logger.py

```diff
@@ -0,0 +1,262 @@
+import logging
+import time
+import uuid
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from autogen import Agent, ConversableAgent
+from autogen.logger.base_logger import BaseLogger
+from openai.types.chat import ChatCompletion
+
+from mlflow import MlflowClient
+from mlflow.entities.span import NoOpSpan, Span, SpanType
+from mlflow.entities.span_event import SpanEvent
+from mlflow.entities.span_status import SpanStatus, SpanStatusCode
+from mlflow.tracing.utils import capture_function_input_args
+
+# For GroupChat, a single "received_message" events are passed around multiple
+# internal layers and thus too verbose if we show them all. Therefore we ignore
+# some of the message senders listed below.
+_EXCLUDED_MESSAGE_SENDERS = ["chat_manager", "checking_agent"]
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ChatState:
+    """
+    Represents the state of a chat session.
+    """
+
+    # The root span object that scopes the entire single chat session. All spans
+    # such as LLM, function calls, in the chat session should be children of this span.
+    session_span: Optional[Span] = None
+    # The last message object in the chat session.
+    last_message: Optional[Any] = None
+    # The timestamp (ns) of the last message in the chat session.
+    last_message_timestamp: int = 0
+    # LLM/Tool Spans created after the last message in the chat session.
+    # We consider them as operations for generating the next message and
+    # re-locate them under the corresponding message span.
+    pended_spans: List[Span] = field(default_factory=list)
+
+    def clear(self):
+        self.session_span = None
+        self.last_message = None
+        self.last_message_timestamp = 0
+        self.pended_spans = []
+
+
+class MlflowAutogenLogger(BaseLogger):
+    def __init__(self):
+        self._client = MlflowClient()
+        self._chat_state = ChatState()
+
+    def start(self) -> str:
+        return "session_id"
+
+    def log_new_agent(self, agent: ConversableAgent, init_args: Dict[str, Any]) -> None:
+        """
+        This handler is called whenever a new agent instance is created.
+        Here we patch the agent's methods to start and end a trace around its chat session.
+        """
+        if hasattr(agent, "initiate_chat"):
+            original = agent.initiate_chat
+            # Setting root_only = True because sometimes compounded agent calls initiate_chat()
+            # method of its sub-agents, which should not start a new trace.
+            agent.initiate_chat = self._patch_function_call(
+                original, "initiate_chat", root_only=True
+            )
```

Comment:
Doesn't this mutate the `ConversableAgent` instance? If so, we should always be using `safe_patch` for that.
Is this patching 100% necessary, or is there another way to go about this? Ideally, we should minimize patching.

Reply:
> Is this patching 100% necessary, or is there another way to go about this?

Hmm, I don't have a solution to do this without a patching. The critical blocker is that their logging system does not tell when the chat starts and ends. It only provides more lower level signal - when a message is generated, LLM is called, etc. There is no ID for it either. Therefore, we need to know boundary of the single "chat" somehow.

AgentOps tackled this limitation by a different approach, which is asking users to manually start and end the "session" ([ref](https://docs.agentops.ai/v1/integrations/autogen)). However, this does not seem to be an option for "autologging".

```
import agentops
agentops.init(<INSERT YOUR API KEY HERE>)
...
# MUST END SESSION at end of program (e.g. main.py)
agentops.end_session("Success") # Success|Fail|Indeterminate
```

Reply:
I can add this to the [feature request](https://github.com/microsoft/autogen/issues/3325) in their repo. Hopefully we can throw away the patching (and the brittle span pending logic) once this FR is resolved.